add NVP implementation

AlejandroCatalina opened the pull request:
This is an initial implementation of real NVP normalising flows.

I didn't implement log_density for the NeuralNet because I only use them as building blocks for the NVP, although they could also be potentially used as BNNs approximating families.

Let me know if there is anything missing here!

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+    def _pth_moment(self, var_param, p):
+        pass
+
+    def supports_pth_moment(self, p):
```
should return `False`

jhuggins replied:
should return `False`

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+    def _pth_moment(self, var_param, p):
+        pass
+
+    def supports_pth_moment(self, p):
```
should return `False`

jhuggins replied:
should return `False`

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+        z = self.forward(x)
+        return z
+
+    def log_density(self, var_param, x):
```
Please add implementation

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+        z = self.forward(x)
+        return z
+
+    def log_density(self, var_param, x):
```
please add implementation

AlejandroCatalina replied:
I'm implementing this but got into a bit of a wall. Because the BNN is modelling an implicit distribution I think we can just evaluate the prior over the variational parameters? There's some work regarding functional BNNs which could be more flexible as they don't impose a prior on the weights themselves but on the function space https://openreview.net/pdf?id=rkxacs0qY7.

jhuggins replied:
I don't know what you mean by prior. If f_lambda is the neural net, then the r.v. is Z = f_lambda(Z_0) for Z_0 ~ N(0, I), so you can compute the density of Z as long as the Jacobian of f_lambda is available, which it is for NFs.

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+        return ldet_mean
+
+    def mean_and_cov(self, var_param):
+        samples = self.sample(var_param, 200000)
```
make number of MC samples an option in the constructor and set default to something smaller (maybe 10K?)

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+        pass
+
+    def mean_and_cov(self, var_param):
+        samples = self.sample(var_param, 200000)
```
make number of MC samples an option in the constructor and set default to something smaller (maybe 10K?)

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+
+class NeuralNet(ApproximationFamily):
+    def __init__(self, layers_shapes, nonlinearity = np.tanh):
+        self._pattern = PatternDict(free_default = True)
```
Need to call `super().__init__(...)`

jhuggins commented on viabel/approximations.py:
```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
+
+class NVPFlow(ApproximationFamily):
+    def __init__(self, layers_t, layers_s, mask, prior, prior_param, dim,  seed = 1):
+        assert len(layers_t) == len(layers_s)
```
Need to call `super().__init__(...)`

jhuggins reviewed (changes requested):
For both  `NeuralNet` and `NVPFlow`, please:
* Add tests to `viabel/tests/test_approximations.py`
* Add documentation for `__init__` and any non-standard functions (e.g., `f`, `g`, and `forward`)
* Add entries to `docs/sources/api/api.rst`

AlejandroCatalina commented:
Yeah, I meant p_0 which is N(0, 1). I implemented log_density now and added doc strings and test functions. Can you check whether they are enough? I used `jacobian` inside log_density to make it general for any non linear function, although we can also use `elementwise_grad` for every layer application.

Best,
Alejandro
On 8. Mar 2021, 16.09 +0200, Jonathan Huggins <notifications@github.com>, wrote:
> @jhuggins commented on this pull request.
> In viabel/approximations.py:
> > +
> +    def forward(self, var_param, x):
> +        for l in range(self._layers):
> +            W = var_param[str(l)]
> +            x = np.dot(x, W)
> +            x = np.where(x < 0, 1e-2 * x, x)
> +        return self._nonlinearity(x)
> +
> +    def sample(self, var_param, n_samples):
> +        x = npr.multivariate_normal(mean = [0] * self.input_dim,
> +                                    cov = np.identity(self.input_dim),
> +                                    size = n_samples)
> +        z = self.forward(x)
> +        return z
> +
> +    def log_density(self, var_param, x):
> I don't know what you mean by prior. If f_lambda is the neural net, then the r.v. is Z = f_lambda(Z_0) for Z_0 ~ N(0, I), so you can compute the density of Z as long as the Jacobian of f_lambda is available, which it is for NFs.
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

AlejandroCatalina commented:
Build seems to be passing now.

AlejandroCatalina commented:
I forgot to add entries to `docs/sources/api/api.rst`, I'll get on with it.
